arrabiata/src/witness.rs

```diff
@@ -0,0 +1,1098 @@
+    /// It is used to allocate new public inputs without having to keep track of
+    /// the position.
+    pub idx_var_pi: usize,
+
+    /// Current processing row. Used to build the witness.
+    pub current_row: usize,
+
+    /// State of the current row in the execution trace
+    pub state: [BigInt; NUMBER_OF_COLUMNS],
+
+    /// Contain the public state
+    // FIXME: I don't like this design. Feel free to suggest a better solution
+    pub public_state: [BigInt; NUMBER_OF_PUBLIC_INPUTS],
+
+    /// Selectors to activate the gadgets.
+    /// The size of the outer vector must be equal to the number of gadgets in
+    /// the circuit.
+    /// The size of the inner vector must be equal to the number of rows in
+    /// the circuit.
+    ///
+    /// The layout columns/rows is used to avoid rebuilding the arrays per
+    /// column when committing to the witness.
+    pub selectors: Vec<Vec<bool>>,
+
+    /// While folding, we must keep track of the challenges the verifier would
+    /// have sent in the SNARK, and we must aggregate them.
+    // FIXME: nothing is done yet, and the challenges haven't been decided yet.
+    // See top-level documentation of the interpreter for more information.
+    pub challenges: Vec<BigInt>,
+
+    /// Keep the current executed instruction
+    /// This can be used to identify which gadget the interpreter is currently
+    /// building.
+    pub current_instruction: Instruction,
+
+    /// The sponges will be used to simulate the verifier messages, and will
+    /// also be used to verify the consistency of the computation by hashing the
+    /// public IO.
+    // IMPROVEME: use a list of BigInt? It might be faster as the CPU will
+    // already have in its cache the values, and we can use a flat array
+    pub sponge_e1: [BigInt; POSEIDON_STATE_SIZE],
+    pub sponge_e2: [BigInt; POSEIDON_STATE_SIZE],
+
+    /// List of public inputs, used first to verify the consistency of the
+    /// previous iteration.
+    pub current_iteration: u64,
+
+    /// A previous hash, encoded in 2 chunks of 128 bits.
+    pub previous_hash: [u128; 2],
+
+    /// The coin folding combiner will be used to generate the combinaison of
+    /// folding instances
+    pub r: BigInt,
+
+    /// Temporary registers for elliptic curve points in affine coordinates than
+    /// can be used to save values between instructions.
+    ///
+    /// These temporary registers can be loaded into the state by using the
+    /// function `load_temporary_accumulators`.
+    ///
+    /// The registers can, and must, be cleaned after the gadget is computed.
+    ///
+    /// The values are considered as BigInt, even though we should add some
+    /// type. As we want to apply the KISS method, we tend to avoid adding
+    /// types. We leave this for future work.
+    ///
+    /// Two registers are provided, represented by a tuple for the coordinates
+    /// (x, y).
+    pub temporary_accumulators: ((BigInt, BigInt), (BigInt, BigInt)),
```

Comment:
I would have something more abstract here, the detail of the EC computation should not appear here
Also we should have a name suggesting it is related to EC

Reply:
Yes, I do agree. It will be abstracted by the perm argument later. The temporary accumulators are removed step by step in following PR.